### DISCUSSION TITLE
MQTT with GSM module

### DISCUSSION THREAD

**Discussion by tadejales** [micropython/micropython#14248] [ESP32]
I am trying to use TTGO T-Call board as MQTT client. I will have this board positioned in remote location where WiFi access is not possible. Is there a known existng MQTT library which I could use. I was searching for it but only found [https://github.com/peterhinch/micropython-mqtt/tree/master](url) which to my knowlage is only avaliable with WiFi interface.

I tought it was worth asking if there is a library in the wild or if @peterhinch work can be customized for this need/used with GSM module instead of Wifi.

Thank you upfront!

**Comment by glenn20** [comment]
You can install a simple mqtt library from [`micropython-lib`](https://github.com/micropython/micropython-lib) using 
```py
import mip
mip.install("mqtt.simple")
```
There is also a variant called "mqtt.robust" - read [the docs](https://github.com/micropython/micropython-lib/tree/master/micropython/umqtt.robust) to see which is right for you.

These libraries assume the device is network connected and don't try to re-establish a wifi connection if network connectivity is lost (as Peter's excellent libraries do).

**Reply by tadejales** [reply]
Thank you. I have found that in the mean time. The problem that I am facing now is that I am unable to connect to MQTT server(Adafruit) over gprs. I am facing OSError -202 . I have tried with the same credentials with cpython sandbox test and I was succesfull so the problem should be somewhere else.

Print from terminal:
` 
File "umqtt/simple.py", line 67, in connect
OSError: -202
`

My mqtt postman:

```
class MQTT:

    def __init__(self) -> None:
        self.config = config.get_config("mqtt")
        self.client = None

        # Exec on init
        self.connect()

    def connect(self) -> None:
        self.client = MQTTClient(client_id=self.config["client_id"], 
                                 server=self.config["broker"], 
                                 port=self.config["port"], 
                                 user=self.config["username"], 
                                 password=self.config["password"])
        
        self.client.connect()
```

But if I understand it right I should be using gsm modem to transmit data.

**Comment by Tico06** [comment]
Hi @tadejales,

I had a similar issue, I need to leave a node in an isolate area. I found a USB dongle which is LTE capable. I created a WIFI router with a Raspberry PI 4 I was holding around.

If you go this way, you have to double check the dongle is compatible with the bands you plan to use.

Another option is to go with an LTE/WIFI router.

**Reply by tadejales** [reply]
I do not have that option as I am also constrained by the space. 

Another option that I am thinking of is to use pure http requests and just post data that way to the server...but I am not able to authenticate myself to adafruit server because library that I have for sim800L does not have ability to include headers to post request.

This is the library I am using: [https://github.com/pythings/Drivers](url)

**Comment by tadejales** [comment]
@glenn20 Any idea what might be the problem with gprs and mqtt.simple? I even dont have idea where to look at.

**Reply by Pawelinode** [reply]
Have you established a network connection via PPP ?

**Comment by ThinkTransit** [comment]
mqtt.robust is an excellent library if you have wifi access or can get a stable ppp connection.

Considering the SIM800 module has inbuilt support for MQTT I would suggest trying this.

https://microchip.ua/simcom/2G/Application%20Notes/SIM800%20Series_MQTT_Application%20Note_V1.03.pdf

**Comment by tadejales** [comment]
@Pawelinode , @ThinkTransit I am sorry but I have no idea what ppp connection is. I am new to any operation with GSM module. I am using library mentioned above and that is what I am working with(I followed the example in the repo and it works: [https://github.com/pythings/Drivers](https://github.com/orgs/micropython/discussions/url)).  I would love to use mqtt.robust if possible. 

Can you give me an example what I would need to do to make it work?

**Comment by Pawelinode** [comment]
You have two options:
1. Establish a network connection via PPP. All MQTT features will work similarly to when connected via WiFi.
2. Don't use Micropython libraries for MQTT. Send data using GSM modem AT commands.

**Reply by tadejales** [reply]
Are you able to give me example how I can connect to PPP. I have find some examples but I was unable to sucesfully establish ppp connection.

I am able to send data via API post request to the Adafruit Io server to simulate publish mqtt functionality, but I am not sure hot to simulate subscribe with API calls.

So if possible I would preffer MQTT library.

**Comment by Pawelinode** [comment]
How did you try to connect via PPP ?
Show code.

**Reply by tadejales** [reply]
Sure. This is what I use.

```
        gsm = UART(1, 9600, timeout=1000, rx=MODEM_TX_PIN, tx=MODEM_RX_PIN)
        time.sleep(1)
        gsm.write("at\r\n")
        print(gsm.readline())
        print(gsm.readline())
        time.sleep(1)
        gsm.write("AT+CPIN?\r\n")
        print(gsm.readline())
        print(gsm.readline())
        time.sleep(1)
        #gsm.write("ate0\r\n")
        #print(gsm.readline())
        #print(gsm.readline())
        #time.sleep(1)
        gsm.write("AT+CREG?\r\n")
        print(gsm.readline())
        print(gsm.readline())
        time.sleep(1)
        gsm.write("AT+CNMI=0,0,0,0,0\r\n")
        print(gsm.readline())
        print(gsm.readline())
        time.sleep(1)
        gsm.write('AT+CGDCONT=1,"IP","globul"\r\n')
        print(gsm.readline())
        print(gsm.readline())
        time.sleep(1)
        gsm.write('AT+CGDATA="PPP",1\r\n')
        time.sleep(1)
        print(gsm.readline())
        print(gsm.readline())

        time.sleep(5)

        import network
        
        ppp=network.PPP(gsm)
        ppp.active(True)
        time.sleep(15)

        print(ppp.ifconfig())
```

And when running it it does not connect. 

I am using TTGO T-Call module.

**Comment by Pawelinode** [comment]
After induction:
gsm.write('AT+CGDCONT=1,"IP","globul"\r\n')
send:
gsm.write('ATD*99#\r\n')
The modem should respond 'CONNECT'

time.sleep(5)

import network
ppp=network.PPP(gsm)
ppp.active(True)
ppp.connect(authmode=ppp.AUTH_PAP, username="", password="")
time.sleep(15)
while ppp.isconnected()==False:
        time.sleep(1)
print(ppp.ifconfig())

**Reply by tadejales** [reply]
So if I understand correctly complete code should loomk like this:

```
...
gsm.write("AT+CNMI=0,0,0,0,0\r\n")
print(gsm.readline())
print(gsm.readline())
time.sleep(1)
gsm.write('AT+CGDCONT=1,"IP","globul"\r\n')

gsm.write('ATD*99#\r\n')

print(gsm.readline())
print(gsm.readline())
time.sleep(1)
gsm.write('AT+CGDATA="PPP",1\r\n')
time.sleep(1)
print(gsm.readline())
print(gsm.readline())

time.sleep(5)

import network
...

```

**Comment by Pawelinode** [comment]
After issuing the gsm.write('ATD*99#\r\n') command, another AT command cannot be sent.
Delete gsm.write('AT+CGDATA="PPP",1\r\n')

I don't know if it's enough to call print(gsm.readline()) twice for each AT command.

**Reply by tadejales** [reply]
Managed to make it work. Thank you for all the support!

But another problem arises that I understand even less. I hope you maybe have a clue?
You can see stripped down version of what I am triying to build but the main procedure is clear. 

I want to execute get request from press of a button, but the script fails. I implemented get request just after connection is established and it executes succesfully. I dont have idea what problem might be, because push button operated without any problems until now.  

Script:
```
from lib.primitives.pushbutton import Pushbutton
from machine import Pin, UART
import time
import network
import urequests as requests
import json
import uasyncio as asyncio

class Postman:

    def __init__(self) -> None:
        read_gsm_iter = 5
        gsm = UART(1, 9600, timeout=1000, rx=26, tx=27)
        time.sleep(1)

        gsm.write("at\r\n")
        for _ in range(read_gsm_iter):
                print(gsm.readline())
        time.sleep(1)

        print("---------- COMMAND: AT+CPIN ----------")
        gsm.write("AT+CPIN?\r\n")
        for _ in range(read_gsm_iter):
                print(gsm.readline())
        time.sleep(1)

        print("---------- COMMAND: ate0 ----------")
        gsm.write("ate0\r\n")
        for _ in range(read_gsm_iter):
                print(gsm.readline())
        time.sleep(1)

        print("---------- COMMAND: AT+CREG ----------")
        gsm.write("AT+CREG?\r\n")
        for _ in range(read_gsm_iter):
                print(gsm.readline())
        time.sleep(1)

        print("---------- COMMAND: AT+CNMI ----------")
        gsm.write("AT+CNMI=0,0,0,0,0\r\n")
        for _ in range(read_gsm_iter):
                print(gsm.readline())
        time.sleep(1)

        print("---------- COMMAND: AT+CGDCONT ----------")
        apn = "internet.t-2.net"
        gsm.write(f'AT+CGDCONT=1,"IP","{apn}"\r\n')
        for _ in range(read_gsm_iter):
                print(gsm.readline())
        time.sleep(1)

        print("---------- COMMAND: ATD*99 ----------")
        gsm.write('ATD*99#\r\n')
        for _ in range(read_gsm_iter):
            print(gsm.readline())
        time.sleep(1)

        time.sleep(5)

        print("----------  PPP ---------- ")
        ppp=network.PPP(gsm)
        ppp.active(True)
        ppp.connect(authmode=ppp.AUTH_PAP, username="", password="")
        time.sleep(15)
        while ppp.isconnected()==False:
            time.sleep(1)

        print(ppp.ifconfig())

        # Try GET request
        print("---------- Test GET request ---------- ")
        ntp_api_url = 'http://worldtimeapi.org/api/timezone/Etc/UTC'
        response = requests.get(ntp_api_url)

        print(response.status_code)
        print(json.loads(response.content))

    def test_get_req(self) -> None:
        print("---------- Execute GET request from button press ----------")
        ntp_api_url = 'http://worldtimeapi.org/api/timezone/Etc/UTC'

        response = requests.get(ntp_api_url)
        print(response)
        print(json.loads(response.content))


# ------------------------------ MAIN -------------------------------------------

# Pin definitions.
position_2_pin = Pin(32, Pin.IN, Pin.PULL_UP)

# Buttons
position_2_pb = Pushbutton(position_2_pin)

# Used for ESP-Now
sta = network.WLAN(network.STA_IF)
sta.active(False)
ap = network.WLAN(network.AP_IF)
ap.active(False)
sta.active(True)
while not sta.active():
    time.sleep(0.1)
sta.config(channel=1)

# Create instance
postman = Postman()

async def main():
    position_2_pb.press_func(postman.test_get_req)

# Start the main coroutine.
loop = asyncio.get_event_loop()
loop.create_task(main())  # Schedule the main coroutine to run
loop.run_forever()  # Run main application code
```

Outcome:
```
b'at\r\r\n'
b'OK\r\n'
None
None
None
---------- COMMAND: AT+CPIN ----------
b'AT+CPIN?\r\r\n'
b'+CPIN: READY\r\n'
b'\r\n'
b'OK\r\n'
b'\r\n'
---------- COMMAND: ate0 ----------
b'Call Ready\r\n'
b'ate0\r\r\n'
b'OK\r\n'
b'\r\n'
b'SMS Ready\r\n'
---------- COMMAND: AT+CREG ----------
b'\r\n'
b'+CREG: 0,2\r\n'
b'\r\n'
b'OK\r\n'
None
---------- COMMAND: AT+CNMI ----------
b'\r\n'
b'OK\r\n'
None
None
None
---------- COMMAND: AT+CGDCONT ----------
b'\r\n'
b'OK\r\n'
None
None
None
---------- COMMAND: ATD*99 ----------
b'\r\n'
b'CONNECT\r\n'
None
b'~\xff}#\xc0!}!}!} }2}"}&} }*} } }#}$\xc0#}\'}"}(}"U\x83~'
None
----------  PPP ---------- 
('10.40.39.197', '192.168.254.254', '255.255.255.255', '84.255.209.79')
---------- Test GET request ---------- 
200
{'timezone': 'Etc/UTC', 'unixtime': 1713262836, 'dst_offset': 0, 'day_of_year': 107, 'abbreviation': 'UTC', 'raw_offset': 0, 'client_ip': '188.230.129.172', 'dst_from': None, 'utc_datetime': '2024-04-16T10:20:36.948137+00:00', 'week_number': 16, 'dst': False, 'datetime': '2024-04-16T10:20:36.948137+00:00', 'utc_offset': '+00:00', 'dst_until': None, 'day_of_week': 2}

---------- Execute GET request from button press ----------
Task exception wasn't retrieved
future: <Task> coro= <generator object '_go' at 3f80d6e0>
Traceback (most recent call last):
  File "asyncio/core.py", line 1, in run_until_complete
  File "lib/primitives/pushbutton.py", line 38, in _go
  File "lib/primitives/pushbutton.py", line 50, in _check
  File "lib/primitives/__init__.py", line 22, in launch
  File "main.py", line 119, in test_get_req
  File "requests/__init__.py", line 180, in get
  File "requests/__init__.py", line 91, in request
OSError: [Errno 113] ECONNABORTED
```

Other error that I am getting is the following, once I get firs, once I get second one.

```
---------- Execute GET request from button press ----------
Task exception wasn't retrieved
future: <Task> coro= <generator object '_go' at 3f810360>
Traceback (most recent call last):
  File "asyncio/core.py", line 1, in run_until_complete
  File "lib/primitives/pushbutton.py", line 38, in _go
  File "lib/primitives/pushbutton.py", line 50, in _check
  File "lib/primitives/__init__.py", line 22, in launch
  File "main.py", line 120, in test_get_req
  File "requests/__init__.py", line 180, in get
  File "requests/__init__.py", line 76, in request
OSError: -202
```

**Reply by peterhinch** [reply]
I think you need to issue `response.close()` to close the socket. My guess is that the host has rejected the second connection because the first is still open (`ECONNABORTED` means a rejection by the host).

**Reply by tadejales** [reply]
I think it is not that. I have also tries with only executing from buttom press and the result was the same. 

I am only adding bottom part of the code that I have changed:

```
...
      print("----------  PPP ---------- ")
      ppp=network.PPP(gsm)
      ppp.active(True)
      ppp.connect(authmode=ppp.AUTH_PAP, username="", password="")
      time.sleep(15)
      while ppp.isconnected()==False:
          time.sleep(1)

      print(ppp.ifconfig())

      # Try GET request
      test_get_req("PROCEDURE GET CALL")

def test_get_req(label) -> None:
  print(f"---------- {label} ----------")
  ntp_api_url = 'http://worldtimeapi.org/api/timezone/Etc/UTC'
  try:
      response = requests.get(ntp_api_url)
      print(response.status_code)
      print(json.loads(response.content))
      response.close()
  except Exception as e:
      print(e)

# ------------------------------ MAIN -------------------------------------------

# Pin definitions.
position_2_pin = Pin(32, Pin.IN, Pin.PULL_UP)

# Buttons
position_2_pb = Pushbutton(position_2_pin)

# Create instance
postman = Postman()

async def main():
  position_2_pb.press_func(test_get_req, ("BUTTON PRESS GET CALL",))

# Start the main coroutine.
loop = asyncio.get_event_loop()
loop.create_task(main())  # Schedule the main coroutine to run
loop.run_forever()  # Run main application code
```

Outcome: 
```
----------  PPP ---------- 
('10.40.232.63', '192.168.254.254', '255.255.255.255', '84.255.209.79')
---------- PROCEDURE GET CALL ----------
200
{'timezone': 'Etc/UTC', 'unixtime': 1713337960, 'dst_offset': 0, 'day_of_year': 108, 'abbreviation': 'UTC', 'raw_offset': 0, 'client_ip': '188.230.129.6', 'dst_from': None, 'utc_datetime': '2024-04-17T07:12:40.429138+00:00', 'week_number': 16, 'dst': False, 'datetime': '2024-04-17T07:12:40.429138+00:00', 'utc_offset': '+00:00', 'dst_until': None, 'day_of_week': 3}
---------- BUTTON PRESS GET CALL ----------
[Errno 113] ECONNABORTED
```

My question would be if I can even run WiFi and PPP at the same time as I need WiFi for ESP-Now, or do they have problems running together. I tried disabling WiFi init procedure shown in the previous post but it made no difference.

**Reply by peterhinch** [reply]
I'm afraid I have no experience of PPP.

You could try changing your invocation of `asyncio`. The following is a more usual approach:
```python
async def main():
  # Instantiate asyncio objects after asyncio has started
  position_2_pb = Pushbutton(position_2_pin)
  position_2_pb.press_func(test_get_req, ("BUTTON PRESS GET CALL",))
  while True:  # Don't let asyncio terminate
    await asyncio.sleep(1)  # In a real app you'd do some work here

asyncio.run(main())  # Start the main coroutine, which runs forever.
```
If this doesn't help, consider that `ECONNABORTED` means connection reset by host. So the question is why does the host respond correctly on the first connection, then reject the second. You could try running the same code on a WiFi-connected board to see if it's something related to PPP (this seems unlikely to me, but you never know). You could even run it on a PC (replacing the button press with an asyncio sleep).

Incidentally I would put the `response.close()` in a `finally` clause so it's called even when an exception is thrown. It won't fix this problem, but it is wise to close sockets.

**Reply by tadejales** [reply]
So first point is that if I put `response.close()` in `finally` I get `NameError: local variable referenced before assignment` error. 

And sure, your suggestion about using WiFi should come to my ming earlier. But it works, with WiFi it works both ways, with procedure call and with pushbutton. So then it should be something related to PPP. 

Do you see any potential problem with async approach with pushbuttons, that that might cause problems with PPP?

```
from machine import Pin
import network
import uasyncio as asyncio
import urequests as requests
import json
import time
from lib.primitives.pushbutton import Pushbutton

wifi = network.WLAN(network.STA_IF)
wifi.active(True)
wifi.connect('UUID', 'PASS')
while not wifi.isconnected():  # Wait until connected...
    time.sleep(0.1)
wifi.status()

# Pin definitions.
position_2_pin = Pin(32, Pin.IN, Pin.PULL_UP)

def test_get_req(label) -> None:
    print(f"---------- {label} ----------")
    ntp_api_url = 'http://worldtimeapi.org/api/timezone/Etc/UTC'
    try:
        response = requests.get(ntp_api_url)
        print(response.status_code)
        print(json.loads(response.content))
        response.close()
    except Exception as e:
        print(e)

test_get_req("PROCEDURE GET CALL")

async def main():
    # Buttons
    position_2_pb = Pushbutton(position_2_pin)
    position_2_pb.press_func(test_get_req, ("BUTTON PRESS GET CALL",))
    while True:
        await asyncio.sleep(1)

asyncio.run(main())
```
Output:
```
---------- PROCEDURE GET CALL ----------
200
{'timezone': 'Etc/UTC', 'unixtime': 1713341521, 'dst_offset': 0, 'day_of_year': 108, 'abbreviation': 'UTC', 'raw_offset': 0, 'client_ip': '89.212.243.251', 'dst_from': None, 'utc_datetime': '2024-04-17T08:12:01.798131+00:00', 'week_number': 16, 'dst': False, 'datetime': '2024-04-17T08:12:01.798131+00:00', 'utc_offset': '+00:00', 'dst_until': None, 'day_of_week': 3}
---------- BUTTON PRESS GET CALL ----------
200
{'timezone': 'Etc/UTC', 'unixtime': 1713341542, 'dst_offset': 0, 'day_of_year': 108, 'abbreviation': 'UTC', 'raw_offset': 0, 'client_ip': '89.212.243.251', 'dst_from': None, 'utc_datetime': '2024-04-17T08:12:22.346011+00:00', 'week_number': 16, 'dst': False, 'datetime': '2024-04-17T08:12:22.346011+00:00', 'utc_offset': '+00:00', 'dst_until': None, 'day_of_week': 3}
```

**Reply by peterhinch** [reply]
Is this a problem with objects going out of scope? You define the `ppp` and `UART` instance in the class constructor, but the scope of those objects is limited to the constructor. I would try making these objects bound variables and ensuring that `test_get_req` is a method.

**Comment by Pawelinode** [comment]
You set a low UART speed and did not define the RTS/CTS line in the modem. There will be a problem with downloading more than ~1.5kB of data (TCP MTU). PPP does not provide flow control. Micropython may lose data sent from the modem.

**Reply by tadejales** [reply]
Can you show me an example. I dont know what you mean by RTS/CTS. By speed you probably mean baudrate? 

For your reference, I am using TTGO T-Call v1.3 board.

**Comment by Pawelinode** [comment]
Modify this line of code:
gsm = UART(1, 9600, timeout=1000, rx=MODEM_TX_PIN, tx=MODEM_RX_PIN)
9600 bps is very low. To communicate with the modem, use the RxD, TxD, CTS, RTS lines and enable hardware flow control RTS/CTS.

**Reply by tadejales** [reply]
I know which line I have to correct, but I hardly know what this naming(RTS, CTS) mean and how this should be done or for what it is used. 

Are you able to explain a bit more and give me an exact example how this would look like in my case. I have made basicaly zero progres. I am not experienced in this and getting stuck this hard is not very fun 😖

**Comment by Pawelinode** [comment]
The problem that occurred may result from the use of requests and async.

See the board schematic. Modems are usually connected using 4 lines RxD, TxD, RTS, CTS.
In UART.init() you provided only RxD, TxD (rx=MODEM_TX_PIN, tx=MODEM_RX_PIN).

**Reply by tadejales** [reply]
Based on the TTGO T-Call v 1.3 there is no exposed RTS, CTS lines that I am aware of. 

![image](https://github.com/micropython/micropython/assets/22613409/ffdcfdcc-6a1c-4aa9-a60d-e5c8e7d24d32)

I tried with 115200 baudrate and it did not made any difference.

**Reply by chris35469** [reply]
OSError: -202  may be linked to a firmware issue. See note regarding downgrading firmware in #13603
